torch/csrc/jit/ir.cpp

```diff
@@ -875,6 +875,197 @@ Node* Node::insertAfter(Node * n) {
   return this;
 }
 
+bool Node::moveAfterTopologicallyValid(Node* n) {
+  return tryMove(n, MoveSide::AFTER);
```

Comment:
Can't it be `tryMove(n->next(), MoveSide::BEFORE)`? `next()` is always valid because of the circular list. `n->next()` may be the `return_node()` but that is ok as long as the topo_index of that node is set to INT_MAX. This would allow MoveSide to be removed and everything to be implemented in terms of moving before something else.

Reply:
No, because we want to guarantee that `this` is adjacent to `n` the specified of side. If we try to move before `n->next()`, we may move `n`.

As a concrete example, if the dependency graph looks like `this -> n -> o` then `moveBefore(o)` will end up with `this, o, n`, but `moveAfter(n)` will return false.

Reply:
Oh, yes-- that is subtle. A comment will help so we don't accidentally regress that.